include/wil/com_apartment_variable.h

```diff
@@ -0,0 +1,230 @@
+    }
+
+    // Get the apartment variable, returning the value stored in the per apartment
+    // storage or use the create function to create and store it in the current apartment.
+    // This variables lifetime is tied to the COM apartment lifetime and will
+    // be safely cleaned up. Call reset_for_current_com_apartment if you want to
+    // rundown the variable before the apartment exits.
+    template<typename F>
+    auto get_for_current_com_apartment(const F& createFn) -> std::invoke_result_t<F>
+    {
+        static_assert(std::is_function_v<F>, "createFn must be a function");
+
+        unsigned long long apartmentId{};
+        FAIL_FAST_IF_FAILED(RoGetApartmentIdentifier(&apartmentId));
+
+        const auto variableKey = std::addressof(createFn);
+
+        auto lock = winrt::slim_lock_guard(details::s_lock);
+
+        auto storage = details::s_apartmentStorage.find(apartmentId);
+        if (storage != details::s_apartmentStorage.end())
+        {
+            auto& variables = storage->second;
+            auto variable = variables.find(variableKey);
+            if (variable != variables.end())
+            {
+                return std::any_cast<std::invoke_result_t<F>>(variable->second);
+            }
+            else
+            {
+                auto result = createFn();
+                variables.insert({ variableKey, result });
+#ifdef _DEBUG
+                variable = variables.find(variableKey);
+                assert(variable != variables.end());
+                std::ignore = std::any_cast<std::invoke_result_t<F>>(variable->second);
+#endif
+                return result;
+            }
+        }
+        else
+        {
+            struct ApartmentObserver : public winrt::implements<ApartmentObserver, IApartmentShutdown>
+            {
+                void STDMETHODCALLTYPE OnUninitialize(unsigned long long apartmentId) noexcept override
+                {
+                    // This code runs at apartment rundown so be careful to avoid deadlocks by
+                    // extracting the variables under the lock then release them outside.
+                    auto variables = [apartmentId]()
+                    {
+                        auto lock = winrt::slim_lock_guard(details::s_lock);
+                        return details::s_apartmentStorage.extract(apartmentId);
+                    }();
+                }
+            };
+
+            auto observer = winrt::make<ApartmentObserver>();
+            unsigned long long id{};
+            APARTMENT_SHUTDOWN_REGISTRATION_COOKIE cookie{}; // leaked, COM run down releases all instances
+            winrt::check_hresult(RoRegisterForApartmentShutdown(observer.get(), &id, &cookie));
+            FAIL_FAST_IF(apartmentId != id);
+
+            auto result = createFn();
+            details::s_apartmentStorage.insert({ apartmentId,
+                decltype(details::s_apartmentStorage)::value_type::second_type { { variableKey, result } } });

```

Comment:
Dunno if this is clearer or more confusing:

```cpp
                decltype(details::s_apartmentStorage[0]) { { variableKey, result } } });
```

but do you even have to say the type?

```cpp
details::s_apartmentStorage.insert({ apartmentId, { { variableKey, result } } });
```

Reply:
thanks, I struggled to get this to compile and see now what you suggested works just fine.